Move `ReachabilityUtils` to WordPressShared

mokagio opened the pull request:
**Problem**

As part of #24165, there are certain pieces that need to be moved which depend on `ReachabilityUtils` an helper object that is also shared with the apps targets.

**Solution**

- Rewrites `ReachabilityUtils` and the underlying `ReachabilityAlert` in Swift
- Moves them to WordPressShared

I considered moving it to a dedicated package because I thought it used the actual Reachability library, but that turned out not to be the case.

I considered ditching this utility that does not support injection or testing, but decided against it to focus on #24165 instead.

mokagio committed:
Define protocol for type proving connection availability

mokagio committed:
Port `ReachabilityAlert` to Swift

mokagio committed:
Rewrite `ReachabilityUtils` in Swift

dangermattic commented:
<!--
  0 Errors
  4 Warnings: <code>Modules/Package.swift</c...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          4 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages as appropriate to your project setup (e.g. in Xcode or by running <code>swift package resolve</code>).</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/308">25.9</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mokagio committed:
Conform `WordPressAppDelegate` to `NetworkConnectionAvailablityGetting`

mokagio committed:
Move `presentFromRootViewController` to WordPressShared

mokagio committed:
Fix test plan in WordPressUITests scheme + do not build screenshots

mokagio committed:
Move `ReachabilityUtils` to WordPressShared

mokagio commented on Modules/Sources/WordPressShared/Utility/UIAlertController+Presentation.swift:
```diff
@@ -0,0 +1,17 @@
+import UIKit
+
+@objc public extension UIAlertController {
+    // FIXME: Given we use the custom leafeViewController, this should be called prestenFromLeafViewController
```
```suggestion
    // FIXME: Given we use the custom leafViewController, this should be called presentFromLeafViewController
```

mokagio committed:
Fix typos

crazytonyli commented on Modules/Sources/WordPressShared/Reachability/ReachabilityUtils.swift:
```diff
@@ -0,0 +1,34 @@
+
+    @objc
+    public static func isInternetReachable() -> Bool {
+        (UIApplication.shared as? NetworkConnectionAvailabilityGetting)?.connectionAvailable ?? false
```
Is this supposed to be using the app delegate instance?
```suggestion
        (UIApplication.shared.delegate as? NetworkConnectionAvailabilityGetting)?.connectionAvailable ?? false
```

crazytonyli replied:
I feel like you could extract the `AppDelegate.configureReachability()` part to here too, so that you don't have to indirectly rely on the app delegate instance.

mokagio committed:
Access application `delegate`, not shared application in `ReachabilityUtils`

Co-authored-by: Tony Li <tony.li@automattic.com>

mokagio commented:
> I feel like you could extract the `AppDelegate.configureReachability()` part to here too, so that you don't have to indirectly rely on the app delegate instance.

Thanks @crazytonyli I'll have a look 👀

mokagio committed:
Add Reachability as a WordPressShared dependency

mokagio committed:
Decouple WordPressAppDelegate from Reachability

mokagio committed:
Remove unused `NetworkConnectionAvailablityGetting`

mokagio committed:
Use `noConnectionMessage` in `ReachabilityAlert`

mokagio commented on WordPress/Classes/ViewRelated/Blog/Blog Details/BlogDetailsViewController.m:
```diff
@@ -1653,15 +1652,14 @@ - (NSString *)propertiesStringForSource:(BlogDetailsNavigationSource)source {
-        [self preloadComments];
-        [self preloadMetadata];
-        [self preloadDomains];
+    if ([ReachabilityUtils.internetReachability isReachableViaWiFi] == false) {
```
Notice this accesses `Reachability`. I think we could work on this further and hide the library altogether under `ReachabilityUtils`. That is, this could call a method on `ReachabilityUtils` instead—_law of Demeter_

mokagio replied:
If this makes sense to you, I'll create an issue to address this once #24165 is "done"

crazytonyli replied:
The Reachability library probably can be replaced with `NWPath`, but that's perhaps a story for another day.

mokagio commented on WordPress/Classes/ViewRelated/Stats/StatsViewController.m:
```diff
@@ -64,8 +65,7 @@ - (void)viewDidLoad
 
-    WordPressAppDelegate *appDelegate = [WordPressAppDelegate shared];
-    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(reachabilityChanged:) name:kReachabilityChangedNotification object:appDelegate.internetReachability];
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(reachabilityChanged:) name:kReachabilityChangedNotification object:ReachabilityUtils.internetReachability];
```
I converted the code here to pass the same `Reachability` instance, but I think we could omit it.

```objc
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(reachabilityChanged:) name:kReachabilityChangedNotification];
```

From [the docs](https://developer.apple.com/documentation/foundation/notificationcenter/1415360-addobserver):

> The object that sends notifications to the observer. Specify a notification sender to deliver only notifications from this sender.
>
> When nil, the notification center doesn’t use sender names as criteria for delivery.

The handler method does

```objc
- (void)reachabilityChanged:(NSNotification *)notification
{
    Reachability *reachability = notification.object;
    if (reachability.isReachable) {
        [self initStats];
    }
}
```

We could replace add the info about connectivity as `userInfo` in the `Notification` and then we'd be good.

mokagio commented on WordPress/WordPress.xcodeproj/project.pbxproj:
```diff
@@ -3000,27 +3000,34 @@
 		};
-		24CE68B42CD3375300C7B37D /* Exceptions for "Classes" folder in "WordPressUITests" target */ = {
+		24CE68AF2CD3375300C7B37D /* Exceptions for "Classes" folder in "WordPressShareExtension" target */ = {
 			isa = PBXFileSystemSynchronizedBuildFileExceptionSet;
 			membershipExceptions = (
-				"Extensions/UIApplication+MainWindow.swift",
+				"Utility/App Configuration/AppConfiguration.swift",
```
Mmm... this and the rest of the diff look like a bad rebase

mokagio committed:
Move `UIApplication+MainWindow` to WordPressShared

mokagio committed:
Use `ReachabilityUtils` in tests over `TestingAppDelegate`

This is not a design choice but a necessity as we now use
`ReachabilityUtils` without passing through the application delegate to
check connectivity.
